### PR TITLE
Support multiple kubeconfig files via env var 'KUBECONFIG'

### PR DESCRIPTION
**Pull request by electricjesus**
Addresses projectcalico/calico#4312

## Description
Prior to this fix, calicoctl does not recognise multiple kubeconfig files specified via KUBECONFIG env var.

This changeset provides more or less copied behaviour of kubectl where it uses `filepath.Split()` and `ClientConfigLoadingRules.Precedence` if multiple kubeconfig targets do exist. 

Falls back to current `ExplicitPath` behaviour setting if config typically set as a single file specified only.

## Todos
- [x] Tests
- [x] Documentation
- [x] Release note

## Release Note

```release-note
`calicoctl` now supports multiple kubeconfig files via KUBECONFIG env var, just as `kubectl` would:

Example  usage:

    export KUBECONFIG=~/.kube/config:~/.kube/kubeconfig_2
    
    calicoctl get node
    calicoctl --context config2 get node
```

### CODE REVIEW

File: lib/backend/k8s/k8s.go
Comment (matthewdupre) [inline lib/backend/k8s/k8s.go:229]: Looks sensible.  I guess the main reason to copy this from client-go's NewDefaultClientConfigLoadingRules() is to maintain our slightly different environment variable setup?

Comment (electricjesus) [reply lib/backend/k8s/k8s.go:229]: Indeed. Almost all of our env vars are setup under calicoctl and passed down as `*apiconfig.CalicoAPIConfigSpec` that would have all the env vars and command line arguments already pre-combined

File: lib/backend/k8s/k8s.go
Comment (fasaxc) [inline lib/backend/k8s/k8s.go:218]: You can do
```
for _, v := range s {
   if encountered[v] {...
```
to directly iterate over the values instead of having to index each time.

File: lib/backend/k8s/k8s_client_test.go
Comment (fasaxc) [inline lib/backend/k8s/k8s_client_test.go:26]: No need to change this because you've only got one test in each block and it would just add a lot of boilerplate(!) but once you've got more than one test in a When/Describe/Context block, it should look like this:
```
When/Describe/Context("there are multiple foos", func() {
  var foos []foo
  var theThing thing
  BeforeEach(func() {
    foos = makeTheFoos()
    theThing = thing.New(thing.WithFoos(foos))
  })
  It("should do bar", func() {
    Expect(theThing).To(DoBar())
  })
  It("should do biff", func() {
    Expect(theThing).To(DoBiff())
  })
})
```
i.e. where you have `When("xyz", func()...)`, "xyz" should be done by the `BeforeEach` block nested inside the `When` so that hte set-up is shared between all the `It`s

File: lib/backend/k8s/k8s.go
Comment (electricjesus) [reply lib/backend/k8s/k8s.go:218]: I think `v` would be a copy now instead of s[i], which would be a reference.. yeah it's weird for such a small func :smile: 

I just pirated this function from https://github.com/kubernetes/client-go/blob/master/tools/clientcmd/loader.go#L635 but we can improve on it certainly

File: lib/backend/k8s/k8s_client_test.go
Comment (electricjesus) [reply lib/backend/k8s/k8s_client_test.go:26]: Right, forgot about BeforeEach! I'll just change it, seems to be structurally better even if it has a lot of boilerplate

File: lib/backend/k8s/k8s.go
Comment (fasaxc) [reply lib/backend/k8s/k8s.go:218]: `for _, v` is semantically a copy, sure, but in practice I'd expect zero difference here; string is a pointer-like type and the compiler is likely to spot that it doesn't need to actually copy.  Also not like this is hot-path critical code!

If you took that code from the k8s repo, please stick a copyright header on it as required by the Apache license.

Comment (electricjesus) [reply lib/backend/k8s/k8s.go:220]: Added the copyright, I just left the function mostly intact

**Review by fasaxc — APPROVED** [review]

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and leave the branch after I've merged it.

**Pull request merged**
